File: 3.longest-substring-without-repeating-characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        count = {}
        start = 0
        max_len = 0
        for i in range(len(s)):
            c = count.get(s[i])
            if c is None or c == 0:
                count[s[i]] = 1
                if i - start + 1 > max_len:
                    max_len = i - start + 1
            else:
                count[s[i]] += 1
                while True:
                    ch = s[start]
                    count[ch] -= 1
                    start += 1
                    if ch == s[i]:
                        break
        return max_len

File: 3.longest-substring-without-repeating-characters/solution.kt
class Solution {
    fun lengthOfLongestSubstring(s: String): Int {
        val count = IntArray(256)
        var maxLen = 0
        var i = 0
        for (j in s.indices) {
            val c = s[j].toInt()
            if (count[c]++ == 0) {
                maxLen = (j - i + 1).coerceAtLeast(maxLen)
            } else {
                var c1: Int
                do {
                    c1 = s[i++].toInt()
                    count[c1]--
                } while (c1 != c)
            }
        }
        return maxLen
    }
}
